sky/data/storage.py

```diff
@@ -3547,3 +3607,450 @@ def _delete_cos_bucket(self):
+            create_dirs: If the local_path is a directory and this is set to
+                False, the contents of the directory are directly uploaded to
+                root of the bucket. If the local_path is a directory and this is
+                set to True, the directory is created in the bucket root and
+                contents are uploaded to it.
+        """
+
+        def get_file_sync_command(base_dir_path, file_names):
+            includes = ' '.join(
+                [f'--include "{file_name}"' for file_name in file_names])
+            sync_command = (
+                'oci os object bulk-upload --no-follow-symlinks --overwrite '
+                f'--bucket-name {self.name} --namespace-name {self.namespace} '
+                f'--src-dir "{base_dir_path}" {includes}')
+
+            all_commands = oci.goto_oci_cli_venv()
+            all_commands.append(sync_command)
+            all_commands.append(oci.leave_oci_cli_venv())
+
+            return ' && '.join(all_commands)
+
+        def get_dir_sync_command(src_dir_path, dest_dir_name):
+            if dest_dir_name and not str(dest_dir_name).endswith('/'):
+                dest_dir_name = f'{dest_dir_name}/'
+
+            # we exclude .git directory from the sync
+            sync_command = (
+                'oci os object bulk-upload --no-follow-symlinks --overwrite '
+                f'--bucket-name {self.name} --namespace-name {self.namespace} '
+                f'--object-prefix "{dest_dir_name}" --src-dir "{src_dir_path}" '
+                f'--exclude ".git/*" ')
+
+            all_commands = oci.goto_oci_cli_venv()
+            all_commands.append(sync_command)
+            all_commands.append(oci.leave_oci_cli_venv())
+
+            return ' && '.join(all_commands)
+
+        # Generate message for upload
+        if len(source_path_list) > 1:
+            source_message = f'{len(source_path_list)} paths'
+        else:
+            source_message = source_path_list[0]
+
+        with rich_utils.safe_status(
+                f'[bold cyan]Syncing '
+                f'[green]{source_message}[/] to [green]oci://{self.name}/[/]'):
+            data_utils.parallel_upload(source_path_list,
+                                       get_file_sync_command,
+                                       get_dir_sync_command,
+                                       self.name,
+                                       self._ACCESS_DENIED_MESSAGE,
+                                       create_dirs=create_dirs,
+                                       max_concurrent_uploads=1)
+
+    def _transfer_to_oci(self) -> None:
+        assert isinstance(self.source, str), self.source
+        if self.source.startswith('s3://'):
+            data_transfer.s3_to_oci(self.name, self.name)
+        elif self.source.startswith('gs://'):
+            data_transfer.gcs_to_oci(self.name, self.name)
+        elif self.source.startswith('r2://'):
+            data_transfer.r2_to_oci(self.name, self.name)
```

Comment:
feeling like we should add some errors here instead of silently ignoring.

Reply:
Throws NotImplementedError directly.